mhkit/dolfyn/adp/discharge.py

```diff
@@ -0,0 +1,182 @@
+    water_depth: xarray.DataArray
+        Total water depth measured by the ADCP or other input, in
+        meters. If measured by the ADCP, add the ADCP's depth below
+        the surface to this array.
+        The "down" direction should be positive.
+    rho: float
+        Water density in kg/m^3
+    mu: float
+        Kinematic visocity based on water temperature and salinity, in Ns/m^2.
+        Required for Reynolds Number.
+    surface_offset: float
+        Surface level offset due to changes in tidal level, in meters.
+        Default: 0 m.
+    utm_zone: int
+        UTM zone that measurements were acquired in. Map of UTM zones for the
+        contiguous US:
+        https://www.usgs.gov/media/images/mapping-utm-grid-conterminous-48-united-states.
+        Default: 10 (the US west coast).
+
+    Returns
+    -------
+    out: xarray.Dataset
+        Dataset containing the following variables:
+         - `discharge`: (1) volume flux, in m^3/s
+         - `power`: (1) power, in W
+         - `power_density`: (1) power density, in W/m^2
+         - `reynolds_number`: (1) Reynolds number, unitless
+    """
+
+    def extrap2bottom(vel, bottom, rng):
+        for idx in range(vel.shape[-1]):
+            z_bot = bottom[idx]
+            # Fetch lowest range index
+            ind_bot = np.nonzero(rng > z_bot)[0][0]
+            for idim in range(vel.shape[0]):
+                vnow = vel[idim, :, idx]
+                # Check that data exists in slice
+                gd = np.isfinite(vnow) & (vnow != 0)
+                if not gd.sum():
+                    continue
+                else:
+                    ind = np.nonzero(gd)[0][-1]
+                z_top = rng[ind]
+                # linearly interpolate next lowest range bin based on 0 m/s at bottom
+                vals = np.interp(rng[ind:ind_bot], [z_top, z_bot], [vnow[ind], 0])
+                vel[idim, ind:ind_bot, idx] = vals
+
+        return vel
+
+    def latlon2utm(ds, proj):
+        PlateC = ccrs.PlateCarree()
+        proj.x0, proj.y0 = proj.transform_point(
+            ds["longitude_gps"].mean(), ds["latitude_gps"].mean(), PlateC
+        )
+        xy = xr.DataArray(
+            proj.transform_points(PlateC, ds["longitude_gps"], ds["latitude_gps"])[
+                :, :2
+            ].T,
+            coords={"gps": ["x", "y"], "time_gps": ds["longitude_gps"]["time_gps"]},
+        )
+
+        # this seems to work for missing latlon
+        xy = xy.interp(
+            time_gps=ds["time"], kwargs={"fill_value": "extrapolate"}
+        ).drop_vars("time_gps")
+        return xy
+
+    def _distance(proj, x, y):
+        # GPS distance traveled in meters
+        return np.sqrt((proj.x0 - x) ** 2 + (proj.y0 - y) ** 2)
+
+    def calc_Q(vel, x, depth, surface_zoff=None):
+        # depth and surface_zoff should be positive in down direction
+        vel = vel.copy()
+        vel = vel.fillna(0)
```

Comment:
Would linear or nearest-neighbor interpolation be more representative than forcing it to zero?

Reply:
Setting it to zero allows you to integrate between it and the next lowest bin that contains data.